dotnet-interactive/resources/code-mirror-lsp.js

```diff
@@ -0,0 +1,154 @@
+                let foundLine = false;
+                let foundColumn = false;
+                let lineNumber = 0;
+                for (; lineNumber < lines.length; lineNumber++) {
+                    let rect = lines[lineNumber].getBoundingClientRect();
+                    if (event.pageX >= rect.left &&
+                        event.pageX <= rect.right &&
+                        event.pageY >= rect.top &&
+                        event.pageY <= rect.bottom) {
+                        foundLine = true;
+                        break;
+                    }
+                }
+
+                // find matching column
+                let columnNumber = 0;
+                if (foundLine) {
+                    let childNodes = lines[lineNumber].children[0].childNodes;
+                    for (let childNode of childNodes) {
+                        let childRect = null;
+                        let childText = "";
+                        if (childNode.nodeType === 1) {
+                            // DOM element
+                            childRect = childNode.getBoundingClientRect();
+                            childText = childNode.innerText;
+                        } else if (childNode.nodeType === 3) {
+                            // text
+                            let range = document.createRange();
+                            range.selectNode(childNode);
+                            childRect = range.getBoundingClientRect();
+                            childText = childNode.textContent;
+                            range.detach();
+                        }
+
+                        if (childRect) {
+                            if (event.pageX >= childRect.left &&
+                                event.pageX <= childRect.right) {
+                                foundColumn = true;
+                                break;
+                            }
+                        }
+
+                        columnNumber += childText.length;
+                    }
+                }
+
+                if (foundColumn) {
+                    global.Lsp.textDocumentHover(element, lineNumber, columnNumber).then(function (result) {
+                        hoverDiv.style.top = `${event.pageY}px`;
+                        hoverDiv.style.left = `${event.pageX}px`;
+                        hoverDiv.style.display = 'inline';
+                        if (result.contents.kind === "markdown") {
+                            require(['components/marked/lib/marked'], function (marked) {
+                                // this should work for jupyter notebook
+                                hoverDiv.innerHTML = marked(result.contents.value);
+                            }, function (_err) {
+                                // couldn't load `marked`; this will happen in jupyter lab
+                                hoverDiv.innerHTML = `<pre>${escapeString(result.contents.value)}</pre>`;
```

Comment:
You should do something like this, which will handle the encoding more correctly than the bespoke `escapeString` method:

```js
let pre = document.createElement('pre');
pre.text = result.contents.value;
hoverDiv.appendChild(pre);
```